### PR TITLE
 executor: fix the behavior when index join meet prefix index

### LINKED ISSUE #11061: Unexcepted result in `INSERT INTO ... SELECT`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
DROP TABLE if exists t1, t2;
CREATE TABLE t1 (c VARCHAR(30), INDEX ix_c (c(10))) charset utf8mb4;
CREATE TABLE t2 (d VARCHAR(10)); 
INSERT INTO t1 (c) VALUES ('7_chars'), ('13_characters'); 

SELECT (SELECT SUM(LENGTH(c)) FROM t1 WHERE c='13_characters') FROM t1;
INSERT INTO t2 (d) 
  SELECT (SELECT SUM(LENGTH(c)) FROM t1 WHERE c='13_characters') FROM t1;

INSERT INTO t2 (d) 
  SELECT (SELECT SUM(LENGTH(c)) FROM t1 WHERE c='7_chars') FROM t1;

ANALYZE table t1, t2;

SELECT (SELECT SUM(LENGTH(c)) FROM t1 WHERE c IN (SELECT t1.c FROM t1)) 
FROM t1;
```

2. What did you expect to see?
```
SELECT * FROM t2 order by d;
d
13
13
20
20
7
7
```

3. What did you see instead?
```
SELECT * FROM t2 order by d;
d
13
13
7
7
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
Release Version: v3.0.0-rc.1-273-gf2841dec9
Git Commit Hash: f2841dec95bf81158aa55a0e26aecb318a3dc208
Git Branch: master
UTC Build Time: 2019-07-03 12:22:23
GoVersion: go version go1.12 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Index join doesn't handle the index with prefix column correctly.
Fix #11061 

### What is changed and how it works?

Cut the column by the length.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: executor/index_lookup_join.go [hunk 1]
```diff
@@ -473,6 +475,10 @@ func (iw *innerWorker) handleTask(ctx context.Context, task *lookUpJoinTask) err
 func (iw *innerWorker) constructLookupContent(task *lookUpJoinTask) ([]*indexJoinLookUpContent, error) {
 	lookUpContents := make([]*indexJoinLookUpContent, 0, task.outerResult.NumRows())
 	keyBuf := make([]byte, 0, 64)
+	keyTypes := make([]*types.FieldType, 0, len(iw.keyCols))
```
Comment (crazycs520) [inline executor/index_lookup_join.go]: move line#478~481 to line #500 ?

Comment (winoros) [reply executor/index_lookup_join.go]: line500 is in a for loop.

Comment (crazycs520) [reply executor/index_lookup_join.go]: before `for loop` and after `if`

File: util/ranger/ranger.go [hunk 2]
```diff
@@ -444,9 +444,9 @@ func fixPrefixColRange(ranges []*Range, lengths []int, tp []*types.FieldType) bo
 	return hasCut
 }
 
-func fixRangeDatum(v *types.Datum, length int, tp *types.FieldType) bool {
-	// If this column is prefix and the prefix length is smaller than the range, cut it.
-	// In case of UTF8, prefix should be cut by characters rather than bytes
+// FixPrefixColDatum cuts the datum according to the prefix length.
+// If it's UTF8 encoded, we will cut it by characters rather than bytes.
+func FixPrefixColDatum(v *types.Datum, length int, tp *types.FieldType) bool {
```
Comment (zz-jason) [inline util/ranger/ranger.go]: how about `s/FixPrefixColDatum/CutPrefix/`?

File: executor/index_lookup_join.go [hunk 3]
```diff
@@ -490,6 +496,16 @@ func (iw *innerWorker) constructLookupContent(task *lookUpJoinTask) ([]*indexJoi
...
 		// Store the encoded lookup key in chunk, so we can use it to lookup the matched inners directly.
 		task.encodedLookUpKeys.AppendBytes(0, keyBuf)
+		if iw.hasPrefixCol {
+			for i := range iw.outerCtx.keyCols {
+				// If it's a prefix column. Try to fix it.
+				if iw.colLens[i] != types.UnspecifiedLength {
+					ranger.FixPrefixColDatum(&dLookUpKey[i], iw.colLens[i], keyTypes[i])
+				}
+			}
+			// dLookUpKey is sorted and deduplicated at sortAndDedupLookUpContents.
```
Comment (zz-jason) [inline executor/index_lookup_join.go]: Unnecessary comment?

File: executor/index_lookup_join.go (hunk 1 above)
Comment (winoros) [reply executor/index_lookup_join.go]: The `if` branch is already in a for loop

File: executor/index_lookup_join.go (hunk 3 above)
Comment (winoros) [reply executor/index_lookup_join.go]: Just to tell why we don't sort it here?

File: executor/index_lookup_join_test.go [hunk 4]
```diff
@@ -152,3 +152,13 @@ func (s *testSuite) TestIndexJoinOverflow(c *C) {
...
 	tk.MustQuery(`select /*+ TIDB_INLJ(t2) */ * from t1 join t2 on t1.a = t2.a;`).Check(testkit.Rows())
 }
+
+func (s *testSuite2) TestIssue11061(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("drop table if exists t1, t2")
+	tk.MustExec("create table t1(c varchar(30), index ix_c(c(10)))")
+	tk.MustExec("create table t2(d varchar(10))")
+	tk.MustExec("insert into t1 (c) values('7_chars'), ('13_characters')")
+	tk.MustExec("insert into t2 (d) values('13'), ('13'), ('7'), ('7')")
```
Comment (lzmhhh123) [inline executor/index_lookup_join_test.go]: Useless table `t2`.

File: executor/index_lookup_join.go [hunk 5]
```diff
@@ -473,6 +475,10 @@ func (iw *innerWorker) handleTask(ctx context.Context, task *lookUpJoinTask) err
 func (iw *innerWorker) constructLookupContent(task *lookUpJoinTask) ([]*indexJoinLookUpContent, error) {
 	lookUpContents := make([]*indexJoinLookUpContent, 0, task.outerResult.NumRows())
 	keyBuf := make([]byte, 0, 64)
+	keyTypes := make([]*types.FieldType, 0, len(iw.keyCols))
+	for _, pos := range iw.keyCols {
+		keyTypes = append(keyTypes, iw.rowTypes[pos])
```
Comment (lzmhhh123) [inline executor/index_lookup_join.go]: What about not saving `keyType`, but use `iw.rowTypes[iw.keyCols[i]]` instead in line 503.

Comment (winoros) [reply executor/index_lookup_join.go]: ok

File: executor/index_lookup_join_test.go (hunk 4 above)
Comment (winoros) [reply executor/index_lookup_join_test.go]: removed

Comment (lzmhhh123) [reply executor/index_lookup_join_test.go]: `insert into t2` isn't removed. 囧

**Pull request merged**